src/libmongoc/src/mongoc/mongoc-shared-private.h

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2021 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mongoc-prelude.h"
+
+#ifndef MONGOC_SHARED_H
+#define MONGOC_SHARED_H
+
+#include <stddef.h>
+
+/**
+ * @brief A ref-counted thread-safe shared pointer to arbitrary data.
+ *
+ * `shared_ptr` instances manage the lifetime of a pointed-to object when the
+ * precise time to destroy that managed object is indeterminate, such as with
+ * shared state in a multithreaded or asynchronous program.
+ *
+ * The pointed-to object of a shared_ptr instance can be accessed via
+ * the `ptr` member of the shared_ptr object. Assigning-to the `ptr` member of
+ * a shared_ptr is legal, and can be done to change the pointed-to object of a
+ * shared_ptr without changing which object is being managed. This can be done
+ * to return a pointer to a subobject of the managed object without giving
+ * out a reference to the full managed object.
+ *
+ * A new managed object with a `shared_ptr` is created with
+ * `mongoc_shared_ptr_create`, which starts with an initial shared reference
+ * count of `1`. To take another reference to keep the managed object alive,
+ * use `mongoc_shared_ptr_copy`. When one is done with a managed resource, the
+ * shared reference should be dropped using `mongoc_shared_ptr_reset_null`.
+ *
+ * When an operation on a shared_ptr causes the reference count to drop to zero,
+ * the destructor that was given to create that shared state will immediately be
```

Comment:
nit: "destructor" here (and in similar places in the documentation and perhaps implementation) should probably be "deleter".

Reply:
Probably correct, as it should both destroy and deallocate (if applicable).

Reply:
In a C++ context, there are still further requirements to be a "destructor", but those wouldn't apply here.

What I'm more trying to highlight is that "deleter" is the terminology used in the Standardese (see util.smartptr.shared).

Reply:
+1, thanks!